doc/man/nsdb-nces.8
.\"@(#)nsdb-nces.8"
.\"
.\" @file doc/man/nsdb-nces.8
.\" @brief man page for nsdb-nces client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH NSDB-NCES 8 "@publication-date@"
.SH NAME
nsdb-nces \- list NSDB container entries on an LDAP server
.SH SYNOPSIS
.B nsdb-nces
.RB [ \-?d ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
The bulk of FedFS junction information in a FedFS domain is stored
on one or more LDAP servers.
These servers are known as
.IR "namespace databases" ,
or NSDBs, for short.
.P
FedFS-enabled file servers and clients access the information stored
on NSDBs via standard LDAP queries.
FedFS-enabled file servers use these queries to resolve FedFS junctions.
FedFS administrators use them to manage information
about file sets contained in a FedFS domain name space.
.SH DESCRIPTION
The
.BR nsdb-nces (8)
command is part of a collection of low-level single-use programs
that in intended for testing the NSDB protocol or for use in scripts.
It queries an LDAP server for the existance of
.IR "NSDB Container Entries" ,
or
.IR NCEs ,
for short.
.P
The top of the Directory Information Tree on an LDAP server has
one or more
.IR "naming contexts" .
Some LDAP server implementations call these contexts "root suffixes".
All LDAP entries on that server are contained under one of these
contexts.
.P
When an LDAP server acts as an NSDB,
one or more LDAP naming contexts on that server contains the
distinguished name of an NCE.
An NCE is the root entry of an LDAP Directory Information Tree
containing FedFS-related information.
This design makes the NCE DN easy for any LDAP client to discover
with no other information than the DNS hostname of the LDAP server.
.P
The
.BR nsdb-nces (8)
command displays each naming context on a target LDAP server
and indicates whether that context contains an NCE.
At its simplest, you can think of the
.BR nsdb-nces (8)
command as a form of NSDB ping.
However, it can also convey certain details about the organization
of any NCEs on an LDAP server.
Discovering NCEs on an NSDB is always the first step
FedFS-enabled file servers perform when resolving a FedFS junction.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Specifies that debugging messages be produced during operation.
.IP "\fB\-?, \-\-help"
Prints an
.BR nsdb-nces (8)
version and usage message on
.IR stderr ,
then exits.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-hostname\fP"
Specifies the hostname of the NSDB to enumerate.
If the
.B --nsdbname
option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B --nsdbname
option is not specified, the
.BR nsdb-nces (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the IP port of the NSDB to enumerate.
If the
.B --nsdbport
option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.SH EXIT CODES
The NSDB returns a value that reflects the success of the requested operation.
.TP
.B FEDFS_OK
The LDAP query succeeded.
One or more NSDB container entries were detected on the target LDAP server.
.TP
.B FEDFS_ERR_ACCESS
The anonymous entity does not have permission to perform the requested operation.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The
.BR nsdb-nces (8)
command was unable to find a route to the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The
.BR nsdb-nces (8)
command determined that the specified NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The
.BR nsdb-nces (8)
command was unable to establish a connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The
.BR nsdb-nces (8)
command was unable to authenticate
and establish a secure connection with the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
A non-specific LDAP error occurred on the connection between the
.BR nsdb-nces (8)
command and specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
An LDAP error occurred on the connection between the
.BR nsdb-nces (8)
command and specified NSDB.
The specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_NONCE
The
.BR nsdb-nces (8)
command was unable to locate any NCEs on the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The
.BR nsdb-nces (8)
command received a malformed response from the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the specified NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The local NSDB connection parameter database
does not have any connection parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The
.BR nsdb-nces (8)
command received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
The
.BR nsdb-nces (8)
command received an LDAP referral that it was unable to follow.
A specific error may be displayed on the command line.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The
.BR nsdb-nces (8)
command received an LDAP referral that it chose not to follow,
either because the local implementation does not support
following LDAP referrals or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The
.BR nsdb-nces (8)
command received an LDAP referral that it chose not to follow
because the local NSDB connection parameter database had no
connection parameters for the NSDB targeted by the LDAP referral.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that you want to know if the LDAP server
.IR ldap.example.net
is an NSDB.  Use:
.RS
.sp
$ nsdb-nces -l ldap.example.net
.br
Host: ldap.example.net:389
.br
  namingContext 'dc=example,dc=net' does not host an NCE.
.br
  namingContext 'o=fedfs' hosts an NCE at 'o=fedfs'.
.br
  namingContext 'o=netscaperoot' does not host an NCE.
.sp
.RE
This shows there are three LDAP naming contexts on the target LDAP server.
One of these is an NSDB Container Entry.
Thus the target LDAP server is an NSDB.
.SH SECURITY
The
.BR nsdb-nces (8)
command uses anonymous binding when performing LDAP queries.
LDAP naming contexts are typically readable by everyone.
.P
The target LDAP server must be registered in the local NSDB connection
parameter database.
The connection security mode listed
in the NSDB connection parameter database
for the target LDAP server is used during this operation.
See
.BR nsdbparams (8)
for details on how to register an NSDB
in the local NSDB connection parameter database.
.SH "SEE ALSO"
.BR fedfs (7),
.BR nsdbparams (8)
.sp
RFC 5716 for FedFS requirements and overview
.sp
RFC 4510 for an introduction to LDAP
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/nsdbc/nsdb-nces.c
/**
 * @file src/nsdbc/nsdb-nces.c
 * @brief List the naming contexts and NCEs on a target NSDB server
 */

/*
 * Copyright 2011 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <netdb.h>
#include <locale.h>
#include <langinfo.h>

#include "fedfs.h"
#include "nsdb.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Short form command line options
 */
static const char nsdb_nces_opts[] = "?de:f:l:r:";

/**
 * Long form command line options
 */
static const struct option nsdb_nces_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nsdbport", 1, NULL, 'r', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 */
static void
nsdb_nces_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [ -d ] "
			"[ -l nsdbname ] [ -r nsdbport ]\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *nsdbname;
	unsigned short nsdbport;
	unsigned int ldap_err;
	FedFsStatus retval;
	char **contexts;
	nsdb_t host;
	int arg, i;

	(void)umask(S_IRWXO);

	/* Ensure UTF-8 strings can be handled transparently */
	if (setlocale(LC_CTYPE, "") == NULL ||
	    strcmp(nl_langinfo(CODESET), "UTF-8") != 0) {
		fprintf(stderr, "Failed to set locale and langinfo\n");
		exit((int)FEDFS_ERR_INVAL);
	}

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb_env(&nsdbname, &nsdbport, NULL, NULL);

	while ((arg = getopt_long(argc, argv, nsdb_nces_opts,
			nsdb_nces_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'l':
			nsdbname = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdbport)) {
				fprintf(stderr, "Bad port number: %s\n",
					optarg);
				nsdb_nces_usage(progname);
			}
			break;
		default:
			fprintf(stderr, "Invalid command line "
				"argument: %c\n", (char)arg);
		case '?':
			nsdb_nces_usage(progname);
		}
	}
	if (optind != argc) {
		fprintf(stderr, "Unrecognized command line argument\n");
		nsdb_nces_usage(progname);
	}
	if (nsdbname == NULL) {
		fprintf(stderr, "Missing required command line argument\n");
		nsdb_nces_usage(progname);
	}

	retval = nsdb_lookup_nsdb(nsdbname, nsdbport, &host, NULL);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_PARAMS:
		fprintf(stderr, "No connection parameters for NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out;
	default:
		fprintf(stderr, "Failed to look up NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out;
	}

	retval = nsdb_open_nsdb(host, NULL, NULL, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_CONN:
		fprintf(stderr, "Failed to connect to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_AUTH:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u\n",
			nsdbname, nsdbport);
		goto out_free;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "Failed to authenticate to NSDB %s:%u: %s\n",
			nsdbname, nsdbport, ldap_err2string(ldap_err));
		goto out_free;
	default:
		fprintf(stderr, "Failed to bind to NSDB %s:%u: %s\n",
			nsdbname, nsdbport,
			nsdb_display_fedfsstatus(retval));
		goto out_free;
	}

	retval = nsdb_get_naming_contexts_s(host, &contexts, &ldap_err);
	switch (retval) {
	case FEDFS_OK:
		break;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "Failed to list NCEs: %s\n",
			ldap_err2string(ldap_err));
		goto out_close;
	default:
		fprintf(stderr, "Failed to list NCEs: %s\n",
			nsdb_display_fedfsstatus(retval));
		goto out_close;
	}

	retval = FEDFS_ERR_NSDB_NONCE;
	printf("NSDB: %s:%u\n", nsdbname, nsdbport);
	for (i = 0; contexts[i] != NULL; i++) {
		char *dn;

		retval = nsdb_get_nceprefix_s(host, contexts[i], &dn, &ldap_err);
		printf("  namingContext '%s' ", contexts[i]);
		if (retval == FEDFS_OK) {
			printf("hosts an NCE at '%s'.\n", dn);
			free(dn);
		} else
			printf("does not host an NCE.\n");
	}

	nsdb_free_string_array(contexts);

out_close:
	nsdb_close_nsdb(host);

out_free:
	nsdb_free_nsdb(host);

out:
	exit((int)retval);
}
